activesupport/test/logger_test.rb

```diff
@@ -120,14 +121,64 @@ def test_buffer_multibyte
     byte_string.force_encoding("ASCII-8BIT")
     assert byte_string.include?(BYTE_STRING)
   end
-  
+
   def test_silencing_everything_but_errors
     @logger.silence do
       @logger.debug "NOT THERE"
       @logger.error "THIS IS HERE"
     end
-    
+
     assert !@output.string.include?("NOT THERE")
     assert @output.string.include?("THIS IS HERE")
   end
+
+  def test_logger_level_thread_safety
+    @logger.level = Logger::INFO
+    assert_level(Logger::INFO)
+
+    thread_1_latch = ActiveSupport::Concurrency::Latch.new
+    thread_2_latch = ActiveSupport::Concurrency::Latch.new
+
+    threads = (1..2).collect do |thread_number|
+      Thread.new do
+        # force thread 2 to wait until thread 1 is already in with_level
+        thread_2_latch.await if thread_number == 2
+
+        with_level(Logger::ERROR) do
+          assert_level(Logger::ERROR)
+
+          # allow thread 2 to finish but hold thread 1
+          if thread_number == 1
+            thread_2_latch.release
+            thread_1_latch.await
+          end
+        end
+
+        # allow thread 1 to finish
+        thread_1_latch.release if thread_number == 2
+        assert_level(Logger::INFO)
```

Comment:
These two lines should be swapped, this way the check for the `info` logger level is done while `thread 1` is still guaranteed to be in `with_level(:error)`.